src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+            findex_master_key,
+
+            fetch_entries_key: None,
+            fetch_chains_key: None,
+            upsert_entries_key: None,
+            insert_chains_key: None,
+        };
+
+        while let Some(prefix) = bytes.next() {
+            let key = Some(bytes.next_chunk::<16>().map_err(|_| {
+                FindexErr::Other(format!(
+                    "the token is too short, expecting 16 bytes after the prefix {prefix}"
+                ))
+            })?);
+
+            if prefix == 0 {
+                token.fetch_entries_key = key;
+            } else if prefix == 1 {
+                token.fetch_chains_key = key;
+            } else if prefix == 2 {
+                token.upsert_entries_key = key;
+            } else if prefix == 3 {
+                token.insert_chains_key = key;
+            } else {
+                return Err(FindexErr::Other(format!(
+                    "the token contains a unknown prefix {prefix}"
+                )));
+            }
+        }
+
+        Ok(token)
+    }
+}
+
+impl Token {
+    pub fn random_findex_master_key(
+        index_id: String,
+        fetch_entries_key: [u8; 16],
+        fetch_chains_key: [u8; 16],
+        upsert_entries_key: [u8; 16],
+        insert_chains_key: [u8; 16],
+    ) -> Result<Self, FindexErr> {
+        let mut rng = CsRng::from_entropy();
+        let findex_master_key = KeyingMaterial::<MASTER_KEY_LENGTH>::new(&mut rng);
+
+        Ok(Token {
+            index_id,
+            findex_master_key,
+            fetch_entries_key: Some(fetch_entries_key),
+            fetch_chains_key: Some(fetch_chains_key),
+            upsert_entries_key: Some(upsert_entries_key),
+            insert_chains_key: Some(insert_chains_key),
+        })
+    }
+
+    pub fn reduce_permissions(&mut self, search: bool, index: bool) -> Result<(), FindexErr> {
+        self.fetch_entries_key =
+            reduce_option("fetch entries", self.fetch_entries_key, search || index)?;
+        self.fetch_chains_key = reduce_option("fetch chains", self.fetch_chains_key, search)?;
+        self.upsert_entries_key = reduce_option("upsert entries", self.upsert_entries_key, index)?;
+        self.insert_chains_key = reduce_option("insert chains", self.insert_chains_key, index)?;
+
+        Ok(())
+    }
+}
+
+fn reduce_option(
+    debug_info: &str,
+    permission: Option<[u8; 16]>,
+    keep: bool,
+) -> Result<Option<[u8; 16]>, FindexErr> {
+    if let Some(permission) = permission {
+        if keep {
+            Ok(Some(permission))
+        } else {
+            Err(FindexErr::Other(format!(
+                "The token provided doesn't have the permission to {debug_info}"
+            )))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+enum Callback {
+    FetchEntries,
+    FetchChains,
+    UpsertEntries,
+    InsertChains,
+}
+
+impl FindexCloud {
+    pub fn new(token: String, base_url: Option<String>) -> Result<Self, FindexErr> {
```

Comment:
I don't understand why you want to use a string representation of your token. It would be much more compact if you use a byte serialization (~30% less by removing the base64 conversion I think?).

Reply:
The token needs to be copy/pastable for a client.

Reply:
Shouldn't this be managed on a higher level ?

Reply:
If you do it client side, you would save allocation space and bandwidth.